consensus/service/src/mint_tx_manager/mod.rs

```diff
@@ -0,0 +1,459 @@
+
+use mc_common::{logger::Logger, HashMap, HashSet};
+use mc_crypto_keys::Ed25519Public;
+use mc_crypto_multisig::SignerSet;
+use mc_ledger_db::Ledger;
+use mc_transaction_core::{
+    mint::{validate_mint_config_tx, MintConfigTx, MintValidationError},
+    BlockVersion, TokenId,
+};
+
+#[derive(Clone)]
+pub struct MintTxManagerImpl<L: Ledger> {
+    /// Ledger DB.
+    ledger_db: L,
+
+    /// The configured block version.
+    block_version: BlockVersion,
+
+    /// A map of token id -> master minters.
+    token_id_to_master_minters: HashMap<TokenId, SignerSet<Ed25519Public>>,
+
+    /// Logger.
+    logger: Logger,
+}
+
+impl<L: Ledger> MintTxManagerImpl<L> {
+    pub fn new(
+        ledger_db: L,
+        block_version: BlockVersion,
+        token_id_to_master_minters: HashMap<TokenId, SignerSet<Ed25519Public>>,
+        logger: Logger,
+    ) -> Self {
+        Self {
+            ledger_db,
+            block_version,
+            token_id_to_master_minters,
+            logger,
+        }
+    }
+}
+
+impl<L: Ledger> MintTxManager for MintTxManagerImpl<L> {
+    /// Validate a MintConfigTx transaction against the current ledger.
+    fn validate_mint_config_tx(&self, mint_config_tx: &MintConfigTx) -> MintTxManagerResult<()> {
+        // Ensure that we have not seen this transaction before.
+        if self
+            .ledger_db
+            .check_mint_config_tx_nonce(&mint_config_tx.prefix.nonce)?
+            .is_some()
+        {
+            return Err(MintTxManagerError::MintValidation(
+                MintValidationError::NonceAlreadyUsed,
+            ));
+        }
+
+        // Get the master minters for this token id.
+        let token_id = TokenId::from(mint_config_tx.prefix.token_id);
+        let master_minters = self.token_id_to_master_minters.get(&token_id).ok_or(
+            MintTxManagerError::MintValidation(MintValidationError::NoMasterMinters(token_id)),
+        )?;
+
+        // Get the current block index.
+        let current_block_index = self.ledger_db.num_blocks()? - 1;
+
+        // Perform the actual validation.
+        validate_mint_config_tx(
+            mint_config_tx,
+            current_block_index,
+            self.block_version,
+            master_minters,
+        )?;
+
+        Ok(())
+    }
+
+    fn combine_mint_config_txs(
+        &self,
+        txs: &[MintConfigTx],
+        max_elements: usize,
+    ) -> MintTxManagerResult<Vec<MintConfigTx>> {
+        let mut seen_nonces = HashSet::default();
+
+        let mut candidates = txs.to_vec();
+        candidates.sort();
+
+        let (allowed_txs, _rejected_txs) = candidates.into_iter().partition(|tx| {
+            if seen_nonces.len() >= max_elements {
+                return false;
+            }
+
+            if seen_nonces.contains(&tx.prefix.nonce) {
+                return false;
+            }
+
+            seen_nonces.insert(tx.prefix.nonce.clone());
+            true
+        });
+
+        Ok(allowed_txs)
```

Comment:
This can be shortened with `Vec::retain` and the return value of `Set::insert`
```suggestion
candidates.retain(|tx| seen_nonces.len() < max_elements && seen_nonces.insert(tx.prefix.nonce.clone());
Ok(candidates)
```

Since we sorted them, we don't even need the set:
```suggestion
        candidates.dedup();
        candidates.truncate(max_elements);
        Ok(candidates)
```

Reply:
... Actually my second suggestion is incorrect, since you were caching the nonces only. There's [`Vec::dedup_by`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.dedup_by) we can use to dedup with just the nonces

Reply:
You know, I think this is actually wrong.
Sorting does not sort by nonce, it sorts by the entire set of fields in the tx.
Lets assume the tx only has amount and nonce, we sort - we might get something like this:
```
(100, nonce1),
(100, nonce2),
(200, nonce3),
(300, nonce1),
```

This will not properly dedupe since the nonces are not consecutive. This can be remedied by sorting by the nonce, but that on its own is not good enough since the sort has to be deterministic. As such, we'll need to sort by nonce + all the other fields. That is messy, and as such I am going to revert this and go back to the HashSet.

Reply:
OK, makes sense